sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1753,15 +1753,26 @@ class DataFrame private[sql](
    * Converts a JavaRDD to a PythonRDD.
    */
   protected[sql] def javaToPython: JavaRDD[Array[Byte]] = {
-    val structType = schema  // capture it for closure
-    val rdd = queryExecution.toRdd.map(EvaluatePython.toJava(_, structType))
-    EvaluatePython.javaToPython(rdd)
+    if (isOutputPickled) {
+      queryExecution.toRdd.map(_.getBinary(0))
```

Comment:
The picked binary can't be used in JVM directly without unpickling, what's the purpose of having non-batched pickled binary format?